_studio/shared/umc/codec/mpeg2_dec/src/umc_mpeg2_decoder.cpp

```diff
@@ -0,0 +1,1147 @@
+        MPEG2DecoderFrameInfo & frameInfo = *frame.GetAU(fieldIndex);
+        frameInfo.SetFilled();
+
+        UpdateDPB(frame, fieldIndex);
+
+        Submit(frame, fieldIndex);
+
+        return UMC::UMC_OK;
+    }
+
+    void MPEG2Decoder::UpdateDPB(MPEG2DecoderFrame& frame, uint8_t fieldIndex)
+    {
+        if (frame.frameType == MPEG2_B_PICTURE || fieldIndex) // Update DPB on reference pictures, the first field for field pictures
+            return;
+
+        uint32_t numRefFrames = std::count_if(m_dpb.begin(), m_dpb.end(),
+                                    [&frame](MPEG2DecoderFrame const * f)
+                                    {
+                                        return (&frame != f) && f->IsRef();
+                                    }
+                                );
+
+        if (NUM_REF_FRAMES == numRefFrames)
+        {
+            auto i = std::min_element(std::begin(m_dpb), std::end(m_dpb),
+                [](MPEG2DecoderFrame const* f1, MPEG2DecoderFrame const* f2)
+                {
+                    uint32_t const id1 = f1->IsRef() ? f1->decOrder : (std::numeric_limits<uint32_t>::max)();
+                    uint32_t const id2 = f2->IsRef() ? f2->decOrder : (std::numeric_limits<uint32_t>::max)();
+
+                    return  id1 < id2;
+                }
+            );
+
+            MPEG2DecoderFrame * oldest = (i == std::end(m_dpb)) ? nullptr : *i;
+            if (oldest)
+                oldest->SetIsRef(false); // Release the oldest reference frame ...
+        }
+
+        frame.SetIsRef(true);            // ... and increase a ref counter for a current frame
+    }
+
+    void MPEG2Decoder::SetDPBSize(uint32_t size)
+    {
+        VM_ASSERT(size > 0);
+        VM_ASSERT(size < 128);
+
+        m_dpbSize = size;
+    }
+
+    void MPEG2Decoder::CompleteDecodedFrames()
+    {
+        std::unique_lock<std::mutex> l(m_guard);
+
+        std::for_each(m_dpb.begin(), m_dpb.end(),
+            [](MPEG2DecoderFrame* frame)
+            {
+                if (frame->IsOutputted() && frame->IsDisplayed() && !frame->IsDecoded())
+                    frame->OnDecodingCompleted();
+            }
+        );
+    }
+
+    MPEG2DecoderFrame* MPEG2Decoder::GetFreeFrame()
+    {
+        MPEG2DecoderFrame* frame = nullptr;
+
+        std::unique_lock<std::mutex> l(m_guard);
+
+        if (m_dpb.size() >= m_dpbSize)
+
+        {
+            auto i = std::find_if(std::begin(m_dpb), std::end(m_dpb),
+                [](MPEG2DecoderFrame const* frame)
+                { return frame->Empty(); }
+            );
+            frame = (i != std::end(m_dpb)) ? *i : nullptr;
+        }
+
+        // Did we find one?
+        if (!frame)
+        {
+            if (m_dpb.size() >= m_dpbSize)
+            {
+                return nullptr;
+            }
+
+            // Didn't find any. Let's create a new one
+            frame = new MPEG2DecoderFrame;
+
+            // Add to DPB
+            m_dpb.push_back(frame);
```

Comment:
it throws, so you actually need to reserve one entry ahead or wrap to smart pointer (first is better I think:) )

Reply:
I didn't get your idea